src/basic/rlimit-util.c

```diff
@@ -76,6 +80,42 @@ int setrlimit_closest_all(const struct rlimit *const *rlim, int *which_failed) {
         return 0;
 }
 
+int setpriority_closest(int nice) {
+        int current, limit;
+        struct rlimit highest;
+
+        /* Try to set requested nice level */
+        if (setpriority(PRIO_PROCESS, 0, nice) == 0)
+                return 0;
+
+        /* Permission failed */
+        if (!IN_SET(errno, EPERM, EACCES))
+                return -errno;
+
+        errno = 0;
+        current = getpriority(PRIO_PROCESS, 0);
+        if (errno != 0)
+                return -errno;
+
```

Comment:
maybe add a shortcut here:

```c
if (nice == current)
        return 0;
```

Reply:
and also something like this:

```c
if (nice > current)
        return -EPERM; /* Hmm, we'd expect that raising the nice level from our status quo would always work. If it doesn't, then the whole setpriority() systemcall is blocked to us, hence let's propagate EPERM right-away */
```

(or if you want to be entirely correct, instead of propagating EPERM here unconditoinally, save the original error code we saw on the initial setpriority() call, and propagate it here, in case it was EACCES)